### PR TITLE
Support replacing the step image in task bundles

### PR DESCRIPTION
**Pull request by joejstuart**
This change introduces a new parameter called STEPS_IMAGE, which allows you to include a custom image for each step in the bundled tasks. 
https://issues.redhat.com/browse/EC-35

### CODE REVIEW

File: hack/test-shellspec.sh [hunk 1]
```diff
@@ -29,7 +29,7 @@ for CHANGED in "${CHANGED_FILES[@]}"; do
             SPEC_DIRS=("${SPEC_DIRS[@]/${SPEC_DIR}}")
             [[ -n "${GITHUB_ACTIONS:-}" ]] && echo "::group::Shellspec test in ${SPEC_DIR}"
             echo -e "Detected changes in \033[1m${CHANGED_DIR}\033[0m, running tests"
-            PARAMS=(--chdir "${SPEC_DIR}" --shell /usr/bin/bash)
+            PARAMS=(--chdir "${SPEC_DIR}" --shell `which bash`)
```
Comment (lcarva) [inline hack/test-shellspec.sh:32]: Seems unrelated to the changes?

File: task/tkn-bundle/0.1/tkn-bundle.yaml [hunk 2]
```diff
@@ -32,17 +36,17 @@ spec:
     env:
     - name: HOME
       value: "$(params.HOME)"
+    - name: TASK_FILE
+      value: tekton_task_files
   steps:
-  - image: quay.io/openshift-pipeline/openshift-pipelines-cli-tkn:5.0
-    name: build
+  - image: quay.io/redhat-appstudio/appstudio-utils:3ef7e414efcb0e625d72ed4b025ea22ab362c60d
```
Comment (lcarva) [inline task/tkn-bundle/0.1/tkn-bundle.yaml:42]: I don't think we should use the `appstudio-utils` image. IIUC its purpose is mostly to assist in the CI of this repo. It is not meant to be used by folks building images in Konflux.

Presumably, you grabbed this image because it's a convenient way to have an image with yq installed. I don't have a good alternative other than maybe just re-use the `quay.io/redhat-appstudio/hacbs-test` image used in the [fbc-related-image-check](https://github.com/redhat-appstudio/build-definitions/blob/3ef7e414efcb0e625d72ed4b025ea22ab362c60d/task/fbc-related-image-check/0.1/fbc-related-image-check.yaml#L20) task. But maybe check with the integration folks that this is ok.

File: task/tkn-bundle/0.1/tkn-bundle.yaml [hunk 3]
```diff
@@ -96,15 +100,42 @@ spec:
           done
         fi
       done
+      for f in "${FILES[@]}"; do
+        if [[ -n "$(params.STEPS_IMAGE)" ]]; then
```
Comment (lcarva) [inline task/tkn-bundle/0.1/tkn-bundle.yaml:104]: If it's not set, maybe don't even enter the for-loop?

File: task/tkn-bundle/0.1/tkn-bundle.yaml [hunk 4]
```diff
@@ -96,15 +100,42 @@ spec:
...
+          yq e '(.spec.steps[] | select(has("image")).image) = "$(params.STEPS_IMAGE)"' -i $f
```
Comment (lcarva) [inline task/tkn-bundle/0.1/tkn-bundle.yaml:105]: It should probably have a way to only modify certain steps. But this is probably good enough for the first iteration. (Not required for this PR.)

File: hack/test-shellspec.sh (hunk 1 above)
Comment (joejstuart) [reply hack/test-shellspec.sh:32]: It is, but I left it here because it allows the script to run on a Mac, which I needed to test the change.

File: task/tkn-bundle/0.1/tkn-bundle.yaml (hunk 2 above)
Comment (joejstuart) [reply task/tkn-bundle/0.1/tkn-bundle.yaml:42]: I was unsure on what to use. Thanks for the suggestion, I'll check with the test team.

Comment (joejstuart) [reply task/tkn-bundle/0.1/tkn-bundle.yaml:42]: actually, shellspec times out when I try and use that image. Not exactly sure why.

File: task/tkn-bundle/0.1/tkn-bundle.yaml (hunk 3 above)
Comment (joejstuart) [reply task/tkn-bundle/0.1/tkn-bundle.yaml:104]: Cleaned it up, should be easier to understand.

File: task/tkn-bundle/0.1/tkn-bundle.yaml [hunk 5]
```diff
@@ -96,15 +100,42 @@ spec:
...
+        fi
+        # filter any duplicate entries and write to workspace file
+        [[ ! `grep "${f}" "${TASK_FILE}"` ]] && echo "$f" >> "${TASK_FILE}"
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply task/tkn-bundle/0.1/tkn-bundle.yaml:108]: This was left over from debugging. I cleaned it up.

File: task/tkn-bundle/0.1/tkn-bundle.yaml [hunk 6]
```diff
@@ -96,15 +100,42 @@ spec:
...
+    - name: IMAGE
+      value: $(params.IMAGE)
+    script: |
+      #!/bin/env bash
+
+      set -o errexit
+      set -o pipefail
+      set -o nounset
+
+      mapfile -t FILES < "${TASK_FILE}"
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply task/tkn-bundle/0.1/tkn-bundle.yaml:125]: Reads the file into an array.

File: hack/test-shellspec.sh (hunk 1 above)
Comment (simonbaird) [reply hack/test-shellspec.sh:32]: where is bash on a mac??

Comment (simonbaird) [reply hack/test-shellspec.sh:32]: Still I'd probably move this up to line 22 or so and maybe comment it out:

```
BASH=/usr/bin/bash
# In case your bash is somewhere else
# BASH=$( which bash )

File: task/tkn-bundle/0.1/tkn-bundle.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (simonbaird) [reply task/tkn-bundle/0.1/tkn-bundle.yaml:126]: But why was it not needed previously?

File: task/tkn-bundle/0.1/tkn-bundle.yaml (hunk 2 above)
Comment (joejstuart) [reply task/tkn-bundle/0.1/tkn-bundle.yaml:42]: I agree that the bundle operations should use that image and I wanted to add the yaml modifications to the existing step, but using sed and awk was messy and looked error prone, so I added the second step for the file operations.

File: task/tkn-bundle/0.1/tkn-bundle.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (joejstuart) [reply task/tkn-bundle/0.1/tkn-bundle.yaml:126]: Because I split the task into multiple steps. The first step passes information using the workspace to the second step. This was needed because the `tkn-cli` image only contained `sed` and `awk` to manipulate the yaml files and I thought that solution was too messy.

File: hack/test-shellspec.sh (hunk 1 above)
Comment (joejstuart) [reply hack/test-shellspec.sh]: What's the benefit of hardcoding the path?

Comment (joejstuart) [reply hack/test-shellspec.sh]: > where is bash on a mac??

If you're using homebrew it's at `/opt/homebrew/bin/bash`

File: task/tkn-bundle/0.1/tkn-bundle.yaml (hunk 2 above)
Comment (joejstuart) [reply task/tkn-bundle/0.1/tkn-bundle.yaml]: Found an alternative image that should be OK. registry.access.redhat.com/ubi9/toolbox@sha256:7391628396216c011ed3a310f1fa54c6a9221e36f7fa59c94ae7796de51e7a25

File: task/tkn-bundle/0.1/tkn-bundle.yaml [hunk 7]
```diff
@@ -32,17 +36,17 @@ spec:
...
+    name: modify-task-files
     env:
     - name: CONTEXT
       value: $(params.CONTEXT)
-    - name: IMAGE
-      value: $(params.IMAGE)
     script: |
       #!/bin/env bash
-
+
```
Comment (zregvart) [inline task/tkn-bundle/0.1/tkn-bundle.yaml:49]: My OCD is triggered by this
```suggestion

```

**Review by zregvart — APPROVED** [review]
Shellspec tests pass, I think we should be okay here.

**Comment by zregvart** [comment]
To get around the yamlint issue change the `on:` to `'on':` in `.github/workflows/check-buildah-remote.yaml` like [here](https://github.com/redhat-appstudio/build-definitions/pull/712/files#diff-96fa6f9702c088c5272e3ea23b828391067207132f9fddd49961ad12cabefbbd). Not sure about the EC check though, I'll try to get the logs before the PipelineRun is GCed.

/retest

**Comment by zregvart** [comment]
We're getting EC policy violations on using disallowed image references in Tasks:

```json
{
  "definitions": [
    {
      "filename": "all_tasks-ec/0.1_build-image-manifest.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_clamav-scan.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.2_init.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_generate-odcs-compose.yaml",
      "violations": [
        {
          "msg": "Step 0 uses disallowed image ref '$(params.IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_source-build.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_git-clone.yaml",
      "violations": [
        {
          "msg": "Step 0 uses disallowed image ref '$(params.gitInitImage)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_prefetch-dependencies.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_fbc-related-image-check.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_buildah-6gb.yaml",
      "violations": [
        {
          "msg": "Step 0 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        },
        {
          "msg": "Step 6 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_tkn-bundle.yaml",
      "violations": [
        {
          "msg": "Step 1 uses disallowed image ref 'quay.io/openshift-pipeline/openshift-pipelines-cli-tkn:5.0'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_s2i-nodejs.yaml",
      "violations": [
        {
          "msg": "Step 1 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        },
        {
          "msg": "Step 4 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_update-infra-deployments.yaml",
      "violations": [
        {
          "msg": "Step 0 uses disallowed image ref '$(params.GIT_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        },
        {
          "msg": "Step 1 uses disallowed image ref '$(params.SCRIPT_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        },
        {
          "msg": "Step 2 uses disallowed image ref '$(params.GIT_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_slack-webhook-notification.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_summary.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_show-sbom.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_buildah-10gb.yaml",
      "violations": [
        {
          "msg": "Step 0 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        },
        {
          "msg": "Step 6 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.3_deprecated-image-check.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_buildah.yaml",
      "violations": [
        {
          "msg": "Step 0 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        },
        {
          "msg": "Step 6 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_s2i-java.yaml",
      "violations": [
        {
          "msg": "Step 1 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        },
        {
          "msg": "Step 5 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_buildah-remote.yaml",
      "violations": [
        {
          "msg": "Step 6 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_rpm-ostree.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_deprecated-image-check.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_clair-scan.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_buildah-8gb.yaml",
      "violations": [
        {
          "msg": "Step 0 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        },
        {
          "msg": "Step 6 uses disallowed image ref '$(params.BUILDER_IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.2_deprecated-image-check.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_sast-snyk-check.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_verify-signed-rpms.yaml",
      "violations": [
        {
          "msg": "Step 0 uses disallowed image ref '$(params.IMAGE)'",
          "metadata": {
            "code": "step_image_registries.step_images_permitted"
          }
        }
      ],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_inspect-image.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_init.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_sbom-json-check.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    },
    {
      "filename": "all_tasks-ec/0.1_fbc-validation.yaml",
      "violations": [],
      "warnings": [],
      "successes": []
    }
  ],
  "success": false,
  "ec-version": "v0.1.2410-2f65d22"
}
```
